aiosmtpd/tests/test_starttls.py

```diff
@@ -0,0 +1,219 @@
+            tls_context=get_tls_context()
+        )
+
+
+class HandshakeFailingHandler:
+    @staticmethod
+    def handle_tls_handshake(*args):
+        return False
+
+
+class TestStartTLS(unittest.TestCase):
+    if _has_ssl:
+        def test_starttls(self):
+            handler = ReceivingHandler()
+            controller = TLSController(handler)
+            controller.start()
+            self.addCleanup(controller.stop)
+            with SMTP(controller.hostname, controller.port) as client:
+                code, response = client.ehlo('example.com')
+                self.assertEqual(code, 250)
+                self.assertIn('starttls', client.esmtp_features)
+                code, response = client.starttls()
+                self.assertEqual(code, 220)
+                client.send_message(
+                    MIMEText('hi'),
+                    'sender@example.com',
+                    'rcpt1@example.com'
+                )
+            self.assertEqual(len(handler.box), 1)
+
+        def test_failed_handshake(self):
+            controller = TLSController(HandshakeFailingHandler())
+            controller.start()
+            self.addCleanup(controller.stop)
+            with SMTP(controller.hostname, controller.port) as client:
+                client.ehlo('example.com')
+                code, response = client.starttls()
+                self.assertEqual(code, 220)
+                code, response = client.mail('sender@example.com')
+                self.assertEqual(code, 554)
+                code, response = client.rcpt('rcpt@example.com')
+                self.assertEqual(code, 554)
+
+    else:
+        def test_starttls_fails_with_no_ssl(self):
+            handler = ReceivingHandler()
+            controller = TLSController(handler)
+            controller.start()
+            self.addCleanup(controller.stop)
+            with SMTP(controller.hostname, controller.port) as client:
+                code, response = client.ehlo('example.com')
+                self.assertEqual(code, 250)
+                self.assertNotIn('starttls', client.esmtp_features)
+                code, response = client.docmd('STARTTLS')
+                self.assertEqual(code, 454)
+
+    def test_disabled_tls(self):
+        controller = Controller(Sink)
+        controller.start()
+        self.addCleanup(controller.stop)
+        with SMTP(controller.hostname, controller.port) as client:
+            client.ehlo('example.com')
+            code, response = client.docmd("STARTTLS")
+            self.assertEqual(code, 454)
+
+    def test_tls_bad_syntax(self):
+        controller = TLSController(Sink)
+        controller.start()
+        self.addCleanup(controller.stop)
+        with SMTP(controller.hostname, controller.port) as client:
+            client.ehlo('example.com')
+            code, response = client.docmd("STARTTLS", "TRUE")
+            self.assertEqual(code, 501)
+
+if _has_ssl:
+    class TestTLSForgetsSessionData(unittest.TestCase):
```

Comment:
Similarly you can `@unittest.skipIf()` for classes too, but do consider whether any of these tests are possible or useful when SSL is not available.